src/EditorFeatures/Text/ITextImageHelpers.cs

```diff
@@ -51,35 +52,54 @@ public static IReadOnlyList<TextChangeRange> GetChangeRanges(ITextImageVersion o
         }
 
         if (changes == null)
-        {
             return [];
-        }
-        else
+
+        var builder = ArrayBuilder<TextChangeRange>.GetInstance(changes.Count);
+        for (var i = 0; i < changes.Count; i++)
         {
-            return ImmutableArray.CreateRange(changes.Select(forward ? s_forwardTextChangeRange : s_backwardTextChangeRange));
+            var change = changes[i];
+            builder.Add(forward ? s_forwardTextChangeRange(change) : s_backwardTextChangeRange(change));
         }
+
+        return builder.ToImmutableAndFree();
     }
 
     private static TextChangeRange GetChangeRanges(ITextImageVersion oldVersion, ITextImageVersion newVersion, bool forward)
     {
         TextChangeRange? range = null;
-        var iterator = GetMultipleVersionTextChanges(oldVersion, newVersion, forward);
-        foreach (var changes in forward ? iterator : iterator.Reverse())
+        using var _ = ArrayBuilder<ArrayBuilder<TextChangeRange>>.GetInstance(out var builder);
+
+        GetMultipleVersionTextChanges(oldVersion, newVersion, forward, builder);
+        foreach (var changes in builder)
         {
             range = range.Accumulate(changes);
+            changes.Free();
         }
 
         RoslynDebug.Assert(range.HasValue);
         return range.Value;
     }
 
-    private static IEnumerable<IEnumerable<TextChangeRange>> GetMultipleVersionTextChanges(
-        ITextImageVersion oldVersion, ITextImageVersion newVersion, bool forward)
+    private static void GetMultipleVersionTextChanges(
+        ITextImageVersion oldVersion,
+        ITextImageVersion newVersion,
+        bool forward,
+        ArrayBuilder<ArrayBuilder<TextChangeRange>> builder)
     {
         for (var version = oldVersion; version != newVersion; version = version.Next)
         {
-            yield return version.Changes.Select(forward ? s_forwardTextChangeRange : s_backwardTextChangeRange);
+            var changes = ArrayBuilder<TextChangeRange>.GetInstance(version.Changes.Count);
```

Comment:
FixedArrayBuilder?

Reply:
Gonna skip on this one as it's easier to just use the ArrayBuilder as it's passed out of the method.